### PR TITLE
Core: Add initialize method for MetricsReporter

### PR DESCRIPTION
**Pull request by wangyinsheng**
A custom MetricsReporter implementation must have a no-arg constructor, but in some case , custom metrics reporter may need args , initialize method will be called after no-arg constructor,  then the reporter can get args through catalog properties

### CODE REVIEW

**Comment by wangyinsheng** [comment]
@nastra will you please take a look this pr ?  And I did not change any code of spark, but  the CI gives errors when check style for spark,  please give some help, thanks

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 1]
```diff
@@ -407,36 +408,38 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
    * @return An initialized {@link MetricsReporter}.
-   * @throws IllegalArgumentException if class path not found or right constructor not found or the
-   *     loaded class cannot be cast to the given interface type
    */
-  public static MetricsReporter loadMetricsReporter(String impl) {
+  public static MetricsReporter loadMetricsReporter(Map<String, String> properties) {
+    String impl = properties.get(CatalogProperties.METRICS_REPORTER_IMPL);
+    if (impl == null || impl.isEmpty()) {
+      return LoggingMetricsReporter.instance();
+    }
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:419]: Style: control flow blocks should have an empty line separating them from the following statement.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 2]
```diff
@@ -407,36 +408,38 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
     LOG.info("Loading custom MetricsReporter implementation: {}", impl);
     DynConstructors.Ctor<MetricsReporter> ctor;
+    MetricsReporter reporter = null;
     try {
       ctor =
           DynConstructors.builder(MetricsReporter.class)
               .loader(CatalogUtil.class.getClassLoader())
               .impl(impl)
               .buildChecked();
+      reporter = ctor.newInstance();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:429]: Please revert the changes that altered how `ClassCastException` is handled.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 3]
```diff
@@ -407,36 +408,38 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
     } catch (NoSuchMethodException e) {
-      throw new IllegalArgumentException(
-          String.format("Cannot initialize MetricsReporter, missing no-arg constructor: %s", impl),
+      LOG.warn(
+          "Cannot initialize MetricsReporter, missing no-arg constructor: {}, fall back to default LoggingMetricsReporter",
+          impl,
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:433]: This should not fall back to logging. It should fail if something is misconfigured.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 4]
```diff
@@ -407,36 +408,38 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
-          String.format(
-              "Cannot initialize MetricsReporter, %s does not implement MetricsReporter.", impl),
+      LOG.warn(
+          "Cannot initialize MetricsReporter, {} does not implement MetricsReporter, fall back to default LoggingMetricsReporter",
+          impl,
           e);
     }
-
+    reporter = reporter == null ? LoggingMetricsReporter.instance() : reporter;
+    reporter.initialize(properties);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:442]: This should be the only addition in this method.

Also, please separate it from the surrounding by newlines since this is a separate task to prepare the reporter.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 5]
```diff
@@ -407,36 +408,38 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
-      throw new IllegalArgumentException(
-          String.format(
-              "Cannot initialize MetricsReporter, %s does not implement MetricsReporter.", impl),
+      LOG.warn(
+          "Cannot initialize MetricsReporter, {} does not implement MetricsReporter, fall back to default LoggingMetricsReporter",
+          impl,
           e);
     }
-
+    reporter = reporter == null ? LoggingMetricsReporter.instance() : reporter;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:441]: This should be unnecessary because the logging reporter is returned at the beginning if the implementation class is not set.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 6]
```diff
@@ -407,36 +408,38 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
+   * @param properties catalog properties which contains class name of a custom {@link
+   *     MetricsReporter} implementation
    * @return An initialized {@link MetricsReporter}.
-   * @throws IllegalArgumentException if class path not found or right constructor not found or the
-   *     loaded class cannot be cast to the given interface type
    */
-  public static MetricsReporter loadMetricsReporter(String impl) {
+  public static MetricsReporter loadMetricsReporter(Map<String, String> properties) {
+    String impl = properties.get(CatalogProperties.METRICS_REPORTER_IMPL);
+    if (impl == null || impl.isEmpty()) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:417]: This changes the behavior and accepts an empty string. I think it should only use the logging reporter if the impl property was not set.

File: core/src/test/java/org/apache/iceberg/TestCatalogUtil.java [hunk 7]
```diff
@@ -193,26 +194,28 @@ public void buildCustomCatalog_withTypeSet() {
...
 
   @Test
   public void loadCustomMetricsReporter_badClass() {
-    Assertions.assertThatThrownBy(
-            () -> CatalogUtil.loadMetricsReporter(TestFileIONotImpl.class.getName()))
-        .isInstanceOf(IllegalArgumentException.class)
-        .hasMessageContaining("does not implement MetricsReporter");
+    Map<String, String> properties = Maps.newHashMap();
+    properties.put(CatalogProperties.METRICS_REPORTER_IMPL, TestFileIONotImpl.class.getName());
+    Assertions.assertThat(CatalogUtil.loadMetricsReporter(properties))
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestCatalogUtil.java:217]: nit: the properties could be inlined: `Assertions.assertThat(CatalogUtil.loadMetricsReporter(ImmutableMap.of(CatalogProperties.METRICS_REPORTER_IMPL, TestFileIONotImpl.class.getName())))`

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 8]
```diff
@@ -407,36 +408,38 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
    *
-   * @param impl full class name of a custom {@link MetricsReporter} implementation
+   * @param properties catalog properties which contains class name of a custom {@link
+   *     MetricsReporter} implementation
    * @return An initialized {@link MetricsReporter}.
-   * @throws IllegalArgumentException if class path not found or right constructor not found or the
-   *     loaded class cannot be cast to the given interface type
    */
-  public static MetricsReporter loadMetricsReporter(String impl) {
+  public static MetricsReporter loadMetricsReporter(Map<String, String> properties) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:415]: I agree with everything what @rdblue mentioned here. 
* it should fallback to the `LoggingMetricsReporter` if there's no reporter available (so that me don't change behavior)
* failures should still happen (as before) if something is misconfigured

**Review by nastra — CHANGES_REQUESTED** [review]
@wangyinsheng thansk for working on this. I've left a few comments.

The CI failure is related to Spark+Checkstyle and we are currently investigating (https://github.com/apache/iceberg/issues/7023) why it happens, so it's not related to your changes.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 1 above)
Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java:420]: done

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 3 above)
Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java]: done

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 4 above)
Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java:447]: done

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 5 above)
Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java]: done

File: core/src/test/java/org/apache/iceberg/TestCatalogUtil.java (hunk 7 above)
Comment (wangyinsheng) [reply core/src/test/java/org/apache/iceberg/TestCatalogUtil.java]: done

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 8 above)
Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java:417]: PTAL

File: api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java [hunk 9]
```diff
@@ -18,10 +18,23 @@
...
+import java.util.Map;
+
 /** This interface defines the basic API for reporting metrics for operations to a Table. */
 @FunctionalInterface
 public interface MetricsReporter {
 
+  /**
+   * Initialize a MetricsReporter with catalog properties.
+   *
+   * <p>A custom MetricsReporter implementation must have a no-arg constructor. After initialize the
```
Comment (nastra) [inline api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java:30]: maybe simplify this to 
```
A custom MetricsReporter implementation must have a no-arg constructor, which will be called first. {@link MetricsReporter#initialize(Map properties)} is called to complete the initialization.
```

**Review by nastra — APPROVED** [review]
This LGTM now, but I would slightly update the javadoc wording

File: api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java [hunk 10]
```diff
@@ -18,10 +18,23 @@
...
+   * MetricsReporter without any arguments, this method should be called to complete MetricsReporter
+   * initialization with catalog properties.
+   *
+   * @param properties catalog properties
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java:34]: These aren't necessarily catalog properties. Just "properties" is fine.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 11]
```diff
@@ -407,12 +408,18 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
-   * @param impl full class name of a custom {@link MetricsReporter} implementation
+   * @param properties catalog properties which contains class name of a custom {@link
+   *     MetricsReporter} implementation
    * @return An initialized {@link MetricsReporter}.
    * @throws IllegalArgumentException if class path not found or right constructor not found or the
    *     loaded class cannot be cast to the given interface type
    */
-  public static MetricsReporter loadMetricsReporter(String impl) {
+  public static MetricsReporter loadMetricsReporter(Map<String, String> properties) {
+    if (!properties.containsKey(CatalogProperties.METRICS_REPORTER_IMPL)) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:418]: I prefer just calling `get` once and checking whether the value is null instead of doing two lookups (`containsKey` followed by `get`).

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 4 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java:447]: There should be an empty line after this to separate it from the return statement.

Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java:447]: done

File: api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java (hunk 10 above)
Comment (wangyinsheng) [reply api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java]: done

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java (hunk 11 above)
Comment (wangyinsheng) [reply core/src/main/java/org/apache/iceberg/CatalogUtil.java]: PTAL

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @wangyinsheng!
